api/src/main/java/com/debugmafia/clueless/model/Board.java

```diff
@@ -0,0 +1,220 @@
+    //TODO: initialize pieces
+    //TODO: initialize weapons
+    shuffleDeck();
+
+  }
+
+  private void shuffleDeck() {
+    List<Card> cardList = new ArrayList<>(deckOCards);
+    Collections.shuffle(cardList);
+    deckOCards = new HashSet<Card>(cardList);
+    hasDeckBeenShuffled = true;
+  }
+
+  public Set<Card> drawWinningCards() {
+    Set<Card> cardsToReturn;
+
+    if (winningCards == null) {
+      cardsToReturn = new HashSet<Card>();
+      if (!hasDeckBeenShuffled) {
+        shuffleDeck();
+      }
+      hasDeckBeenShuffled = true;
+
+      for (Card c : deckOCards) {
+        List<Card> filteredList = cardsToReturn.stream().filter(x -> x.getType() == c.getType())
+            .collect(Collectors.toList());
+        if (filteredList.isEmpty()) {
+          cardsToReturn.add(c);
+          deckOCards.remove(c);
+        }
+      }
+
+      winningCards = cardsToReturn;
+    } else {
+      cardsToReturn = winningCards;
+    }
+    return cardsToReturn;
+  }
+
+  public Set<Card> draw(int numToDraw) {
+
+    HashSet<Card> setToReturn = null;
+
+    if (hasDeckBeenShuffled && deckOCards.size() >= numToDraw) {
+
+      Iterator<Card> iterator = deckOCards.iterator();
+
+      while(numToDraw > 0 && iterator.hasNext())
+      {
+        setToReturn.add(iterator.next());
+        iterator.remove();
+        numToDraw--;
+      }
+    }
+    return setToReturn;
+  }
+
+  public Set<BoardLocation> getOpenAdjacentLocations(BoardLocation locationToQuery) {
+    Set<BoardLocation> validLocations = new HashSet<BoardLocation>();
+
+    for (BoardLocation[] row : grid) {
+      for(BoardLocation column: row)
+      {
+      if (locationToQuery.getXcoord() + 1 == column.getXcoord() || locationToQuery.getXcoord() + 1 == column.getYcoord()) {
+        if (column.getType() == BoardLocationType.ROOM) {
+          validLocations.add(column);
+        } else if (column.getType() == BoardLocationType.HALLWAY && locationToQuery.getPieces().isEmpty()) {
+          validLocations.add(column);
+        }
+      } else if (locationToQuery.getSecretPassage() != null && locationToQuery.getSecretPassage() == column) {
+        validLocations.add(column);
+      }
+    }
+    }
+
+    return validLocations;
+  }
+
+  public Board movePiece(Piece p, BoardLocation to) {
+
+    BoardLocation localLocation = null;
+    Piece localPiece = null;
+
+    for(BoardLocation[] row : grid )    
+    {
+      for(BoardLocation column : row)
+      {
+        if(column == to)
+        {
+          localLocation = column;
+          break;
+        }
+      }
+    }
+
+    for(Piece pIter : gamePieces)
+    {
+      if(pIter == p)
+      {
+        localPiece = pIter;
+        break;
+      }
+    }
+
+    if(localLocation == null || localPiece == null)
+    {
+      return null;
+    }
+
+    HashSet<Piece> localPieces = new HashSet<Piece>(localLocation.getPieces());
+
+    localPieces.add(localPiece);
+    localLocation.setPieces(localPieces);
```

Comment:
After we add the piece to the new location we have to do something else - _remove_ the piece from the old location since it'll still be there.